src/Nethermind/Nethermind.Network.Discovery/RoutingTable/NodeBucket.cs

```diff
@@ -28,26 +29,71 @@ public NodeBucket(int distance, int bucketSize, float dropFullBucketProbability
 
     public int BucketSize { get; }
 
-    public IEnumerable<NodeBucketItem> BondedItems
+    public bool AnyBondedItems()
     {
-        get
+        foreach (NodeBucketItem _ in BondedItems)
         {
-            lock (_nodeBucketLock)
+            return true;
+        }
+
+        return false;
+    }
+
+    public BondedItemsEnumerator BondedItems
+        => new(this);
+
+    public struct BondedItemsEnumerator : IEnumerator<NodeBucketItem>, IEnumerable<NodeBucketItem>
+    {
+        private NodeBucket _nodeBucket;
+        private LinkedListNode<NodeBucketItem>? _currentNode;
+        private DateTime _referenceTime;
+
+        public BondedItemsEnumerator(NodeBucket nodeBucket)
+        {
+            _nodeBucket = nodeBucket;
+            _referenceTime = DateTime.UtcNow;
+            Monitor.Enter(_nodeBucket._nodeBucketLock);
+            _currentNode = nodeBucket._items.Last;
+            Current = null!;
+        }
+
+        public NodeBucketItem Current { get; private set; }
+
+        object IEnumerator.Current => Current;
+
+        public bool MoveNext()
+        {
+            while (_currentNode is not null)
             {
-                LinkedListNode<NodeBucketItem>? node = _items.Last;
-                DateTime utcNow = DateTime.UtcNow;
-                while (node is not null)
+                Current = _currentNode.Value;
+                _currentNode = _currentNode.Previous;
+                if (Current.IsBonded(_referenceTime))
                 {
-                    if (!node.Value.IsBonded(utcNow))
-                    {
-                        break;
-                    }
-
-                    yield return node.Value;
-                    node = node.Previous;
+                    return true;
                 }
             }
+
+            Current = null!;
+            return false;
+        }
+
+        public void Reset() => throw new NotSupportedException();
```

Comment:
Should we have Reset's throw not supported exceptions? Shouldn't be hard to reset all of them?

Reply:
Not if never use it, docs say don't bother https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerator.reset?view=net-8.0#remarks

![image](https://github.com/NethermindEth/nethermind/assets/1142958/5176b280-22a0-4dfd-8b3f-1a1605553eee)


Hid behind explicit implementation https://github.com/NethermindEth/nethermind/pull/6967/commits/e2586909ec940bb9271620645abedd6ddfe058a9